hadoop-hdds/docs/content/design/storage-class.md

```diff
@@ -19,10 +19,331 @@ author: Marton Ele
+It's not possible to defined storage-class (or any replication rule) with using *Hadoop Compatible File System* interface. However storage-class defined on bucket level (or prefix level) will be inherited, even if the keys are created view the `o3fs://` or `o3s://` interfaces
+
+# The storage-class as an abstraction
+
+The previos section explained some user facing property of the storage-class concept. This section explains the concept compared to the existing Ozone design.
+
+## Containers in more details
+
+Container is the unit of replication of Ozone. One Container can store multiple blocks (default container size is 5GB) and they are replicated together. Datanodes report only the replication state of the Containers back to the Storage Container Manager (SCM) which makes it possible to scale up to billions of objects.
+
+The identifier of a block (BlockId) containers ContainerId and LocalId (ID inside the container). ContainerId can be used to find the right Datanode which stores the data. LocalId can be used to find the data inside one container.
+
+Container type defines the following:
+
+ * How to write to the containers?
+ * How to read from the containers?
+ * How to recover / replicate data in case of error
+ * How to store the data on the Datanode (related to the *how to write* question?)
+
+THe current definition of *Ratis/THREE* is the following (simplified version):
+
+ * **How to write**: Call standard Datanode RPC API on *Leader*. Leader will replicate the data to the followers
+ * **How to read**: Read the data from the Leader (stale read can be possible long-term)
+ * **How to replicate / recover**
+    * Transient failures can be handled by new leader election
+    * Permanent degradation couldn't be handled. (Transition to Closed containers is required)
+
+The current definitions of the *Closed/THREE*:
+
+  * **How to write**: Closed containers are not writeable
+  * **How to read**: Read the data from any nodes (Simple RPC call to the DN)
+  * **How to replicate / recover**
+    * Datanodes provides a GRPC endpoint to publish containers as compressed package
+    * Replication Manager (SCM) can send commands to DN to replicate data FROM other Datanode
+    * Datanode downloads the compressed package and import it
+
+The definitions of *Closed/ONE*:
+  * **How to write**: Closed containers are not writeable
+  * **How to read**: Read the data from any nodes (Simple RPC call to the DN)
+  * **How to replicate / recover**: No recovery, sorry.
+
+
+## Storage-class
+
+Let's define the *storage-class* as set of used container **types and transitions** between them during the life cycle of the containers.
+
+The type of the Container can be defined with the implementation type (eg. Ratis, EC, Closed) and with additional parameters related to type (eg. replication type of Ratis, or EC algorithm for EC containers).
+
+Today's implementation of Ozone can be described with two storage-classes:
+
+![](https://i.imgur.com/02rmoFP.png)
```

Comment:
Can we put the png in ozone rather than imgur ? If imgur crash or offline, we can not visit it.

Reply:
Good point. I added both the pictures to make it available offline. (#726843f)